beacon_node/network/src/sync/range_sync/range.rs

```diff
@@ -356,3 +361,257 @@ impl<T: BeaconChainTypes> RangeSync<T> {
+        let chain = Arc::new(FakeStorage::default());
+        let log = build_log(slog::Level::Trace, log_enabled);
+        let (beacon_processor_tx, beacon_processor_rx) = mpsc::channel(10);
+        let range_sync = RangeSync::<TestBeaconChainType, Arc<FakeStorage>>::new(
+            chain.clone(),
+            beacon_processor_tx,
+            log.new(o!("component" => "range")),
+        );
+        let (network_tx, network_rx) = mpsc::unbounded_channel();
+        let globals = {
+            use lighthouse_network::discovery::enr_ext::CombinedKeyExt;
+            use lighthouse_network::discv5::enr::CombinedKey;
+            use lighthouse_network::discv5::enr::EnrBuilder;
+            use lighthouse_network::rpc::methods::{MetaData, MetaDataV2};
+
+            let keypair = libp2p::identity::Keypair::generate_secp256k1();
+            let enr_key: CombinedKey = CombinedKey::from_libp2p(&keypair).unwrap();
+            let enr = EnrBuilder::new("v4").build(&enr_key).unwrap();
+            let globals = NetworkGlobals::new(
+                enr,
+                9000,
+                9000,
+                MetaData::V2(MetaDataV2 {
+                    seq_number: 0,
+                    attnets: Default::default(),
+                    syncnets: Default::default(),
+                }),
+                vec![],
+                &log,
+            );
+            Arc::new(globals)
+        };
+        let cx = SyncNetworkContext::new(
+            network_tx,
+            globals.clone(),
+            log.new(o!("component" => "network_context")),
+        );
+        let test_rig = TestRig {
+            log,
+            beacon_processor_rx,
+            chain,
+            cx,
+            network_rx,
+            globals,
+        };
+        (test_rig, range_sync)
+    }
+
+    #[test]
+    fn head_chain_removed_while_finalized_syncing() {
+        // NOTE: this is a regression test.
+        let (mut rig, mut range) = range(true);
+
+        // Get a peer with an advanced head
+        let (head_peer, local_info, remote_info) = rig.head_peer();
+        range.add_peer(&mut rig.cx, local_info, head_peer, remote_info);
+        range.assert_state(RangeSyncType::Head);
+
+        // Sync should have requested a batch, grab the request.
+        let _request = rig.grab_request(&head_peer);
+
+        // Now get a peer with an advanced finalized epoch.
+        let (finalized_peer, local_info, remote_info) = rig.finalized_peer();
+        range.add_peer(&mut rig.cx, local_info, finalized_peer, remote_info);
+        range.assert_state(RangeSyncType::Finalized);
+
+        // Sync should have requested a batch, grab the request
+        let _second_request = rig.grab_request(&finalized_peer);
+
+        // Fail the head chain by disconnecting the peer.
+        range.remove_peer(&mut rig.cx, &head_peer);
+        range.assert_state(RangeSyncType::Finalized);
```

Comment:
I'm trying to understand what this assertion wanted to prove. Removing the head chain does not affect the sync state and it's already in Finalized type. What the buggy behaviour that this test fixed?

Reply:
hey @dapplion, taking into account this was 3 years ago I most definitely do not remember the details. I restored the branch, if you want you can `gh pr checkout 2821` and revert the couple of assertions that I changed. Patch ahead

```diff
diff --git a/beacon_node/network/src/sync/range_sync/chain_collection.rs b/beacon_node/network/src/sync/range_sync/chain_collection.rs
index 4dc9c1d01..265c8b897 100644
--- a/beacon_node/network/src/sync/range_sync/chain_collection.rs
+++ b/beacon_node/network/src/sync/range_sync/chain_collection.rs
@@ -72,7 +72,7 @@ impl<T: BeaconChainTypes, C: BlockStorage> ChainCollection<T, C> {
             RangeSyncState::Finalized(ref syncing_id) => {
                 if syncing_id == id {
                     // the finalized chain that was syncing was removed
-                    debug_assert!(was_syncing && sync_type == RangeSyncType::Finalized);
+                    debug_assert!(was_syncing);
                     let syncing_head_ids: SmallVec<[u64; PARALLEL_HEAD_CHAINS]> = self
                         .head_chains
                         .iter()
@@ -86,7 +86,7 @@ impl<T: BeaconChainTypes, C: BlockStorage> ChainCollection<T, C> {
                     };
                 } else {
                     // we removed a head chain, or an stoped finalized chain
-                    debug_assert!(!was_syncing || sync_type != RangeSyncType::Finalized);
+                    debug_assert!(!was_syncing);
                 }
             }
             RangeSyncState::Head(ref mut syncing_head_ids) => {
@@ -507,4 +507,4 @@ impl<T: BeaconChainTypes, C: BlockStorage> ChainCollection<T, C> {
             }
         }
     }
-}
+}
\ No newline at end of file
```

and the test fails.

The specific assertion you ask, I don't know. However (at least back then) removing a peer might mean the state changes if it's the only peer on the chain that is syncing, for example. Other reason might have been there as well, but this is what comes to mind.